### PR TITLE
Anko - Update group data source

### PR DESCRIPTION
**Pull request by at-cuongcao**
#### Related issues (Required)
- Task #161 

#### Description (Required)
- [x] Update group data source

### CODE REVIEW

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 1]
```diff
@@ -15,8 +16,8 @@ class GroupRepository : GroupDataSource {
 
     private val remoteDataSource = GroupRemoteDataSource()
 
-    override fun getGroupId(userId: String): Observable<String> {
-        return remoteDataSource.getGroupId(userId)
+    override fun listenerForGroupChange(userId: String): Observable<String> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 2]
```diff
@@ -27,27 +28,59 @@ class GroupRepository : GroupDataSource {
         return remoteDataSource.getInvite(userId)
     }
 
-    override fun getRequest(groupId: String): Observable<Invite> {
-        TODO("not implemented")
+    override fun getGroupRequest(groupId: String): Observable<Invite> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 3]
```diff
@@ -27,27 +28,59 @@ class GroupRepository : GroupDataSource {
...
+        return remoteDataSource.getGroupRequest(groupId)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 4]
```diff
@@ -27,27 +28,59 @@ class GroupRepository : GroupDataSource {
...
     }
 
-    override fun upGroupInfo(group: Group): Observable<Boolean> {
-        TODO("not implemented")
+    override fun postGroupInfo(group: Group): Observable<Boolean> {
+        return remoteDataSource.postGroupInfo(group)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 5]
```diff
@@ -27,27 +28,59 @@ class GroupRepository : GroupDataSource {
...
     }
 
-    override fun changeOwner(groupId: String, newOwner: String): Observable<Boolean> {
-        TODO("not implemented")
+    override fun changeGroupOwner(groupId: String, newOwner: String): Single<Boolean> {
+        return remoteDataSource.changeGroupOwner(groupId, newOwner)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 6]
```diff
@@ -27,27 +28,59 @@ class GroupRepository : GroupDataSource {
...
     }
 
     override fun removeGroup(groupId: String): Observable<Boolean> {
-        TODO("not implemented")
+        return remoteDataSource.removeGroup(groupId)
     }
 
-    override fun upInvite(userId: String, invite: Invite): Single<Boolean> {
-        TODO("not implemented")
+    override fun postInvite(userId: String, invite: Invite): Single<Boolean> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 7]
```diff
@@ -27,27 +28,59 @@ class GroupRepository : GroupDataSource {
...
+        return remoteDataSource.postInvite(userId, invite)
     }
 
     override fun removeUserFromGroup(userId: String): Single<Boolean> {
-        TODO("not implemented")
+        return remoteDataSource.removeUserFromGroup(userId)
+    }
+
+    override fun searchGroup(groupName: String): Observable<List<Group>> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 8]
```diff
@@ -27,27 +28,59 @@ class GroupRepository : GroupDataSource {
...
+        return remoteDataSource.searchGroup(groupName)
+    }
+
+    override fun getCurrentRequestOfUser(userId: String): Observable<Invite> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 9]
```diff
@@ -27,27 +28,59 @@ class GroupRepository : GroupDataSource {
...
+        return remoteDataSource.getCurrentRequestOfUser(userId)
+    }
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 10]
```diff
@@ -27,27 +28,59 @@ class GroupRepository : GroupDataSource {
...
+
+    override fun postRequestToGroup(groupId: String, request: Invite): Single<Boolean> {
+        return remoteDataSource.postRequestToGroup(groupId, request)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 11]
```diff
@@ -27,27 +28,59 @@ class GroupRepository : GroupDataSource {
...
+    }
+
+    override fun postRequestToUser(userId: String, request: Invite): Single<Boolean> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 12]
```diff
@@ -27,27 +28,59 @@ class GroupRepository : GroupDataSource {
...
+        return remoteDataSource.postRequestToGroup(userId, request)
+    }
+
+    override fun searchUser(name: String): Observable<List<User>> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/remote/GroupRemoteDataSource.kt [hunk 13]
```diff
@@ -122,7 +131,118 @@ class GroupRemoteDataSource : GroupDataSource {
...
     override fun removeUserFromGroup(userId: String): Single<Boolean> {
-        TODO("not implemented")
+        val result = SingleSubject.create<Boolean>()
+        HypertrackApi.instance.removeUserFromGroup(userId, BodyAddUserToGroup(null))
+                .doOnSuccess { result.onSuccess(true) }
+                .doOnError { result.onError(it) }
+        return result
+    }
+
+    override fun searchGroup(groupName: String): Observable<List<Group>> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/GroupRemoteDataSource.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/remote/GroupRemoteDataSource.kt [hunk 14]
```diff
@@ -122,7 +131,118 @@ class GroupRemoteDataSource : GroupDataSource {
...
+        val requestRef = firebaseDatabase.getReference("user/$userId/request")
+        requestRef.setValue(request).addOnSuccessListener {
+            result.onSuccess(true)
+        }.addOnFailureListener {
+            result.onError(it)
+        }
+        return result
+    }
+
+    override fun searchUser(name: String): Observable<List<User>> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/remote/GroupRemoteDataSource.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 15]
```diff
@@ -27,27 +28,63 @@ class GroupRepository : GroupDataSource {
...
+        return remoteDataSource.searchUser(name)
+    }
+
+    override fun deleteGroupRequest(groupId: String, request: Invite): Single<Boolean> {
+        return remoteDataSource.deleteGroupRequest(groupId, request)
+    }
+
+    override fun deleteUserInvite(userId: String, invite: Invite): Single<Boolean> {
+        return remoteDataSource.deleteUserInvite(userId, invite)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 16]
```diff
@@ -16,8 +16,8 @@ class GroupRepository : GroupDataSource {
 
     private val remoteDataSource = GroupRemoteDataSource()
 
-    override fun getGroupId(userId: String): Observable<String> {
-        return remoteDataSource.getGroupId(userId)
+    override fun listenerForGroupChange(userId: String): Observable<String> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 17]
```diff
@@ -60,15 +60,31 @@ class GroupRepository : GroupDataSource {
         return remoteDataSource.getCurrentRequestOfUser(userId)
     }
 
-    override fun postRequestToGroup(groupId: String, request: Invite) {
-        remoteDataSource.postRequestToGroup(groupId, request)
+    override fun postRequestToGroup(groupId: String, request: Invite): Single<Boolean> {
+        return remoteDataSource.postRequestToGroup(groupId, request)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 18]
```diff
@@ -60,15 +60,31 @@ class GroupRepository : GroupDataSource {
...
     }
 
-    override fun postRequestToUser(userId: String, request: Invite) {
-        remoteDataSource.postRequestToGroup(userId, request)
+    override fun postRequestToUser(userId: String, request: Invite): Single<Boolean> {
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt [hunk 19]
```diff
@@ -60,15 +60,31 @@ class GroupRepository : GroupDataSource {
...
     override fun searchUser(name: String): Observable<List<User>> {
         return remoteDataSource.searchUser(name)
     }
+
+    override fun deleteGroupRequest(groupId: String, request: Invite): Single<Boolean> {
+        return remoteDataSource.deleteGroupRequest(groupId, request)
+    }
+
+    override fun deleteUserInvite(userId: String, invite: Invite): Single<Boolean> {
+        return remoteDataSource.deleteUserInvite(userId, invite)
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/way/data/source/GroupRepository.kt]: WARNING:

File: app/src/main/java/vn/asiantech/way/data/source/remote/GroupRemoteDataSource.kt [hunk 20]
```diff
@@ -151,28 +155,132 @@ class GroupRemoteDataSource : GroupDataSource {
...
+                                val requestRef = firebaseDatabase.getReference("group/$groupId" +
+                                        "/request/${request.from}")
+                                requestRef.setValue(request).addOnSuccessListener {
+                                    result.onSuccess(true)
+                                }.addOnFailureListener {
+                                    result.onError(it)
+                                }
+                            }
+                            .addOnFailureListener {
+                                result.onError(Throwable(it))
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/way/data/source/remote/GroupRemoteDataSource.kt]: We do not cast to `Throwable`

File: app/src/main/java/vn/asiantech/way/data/source/remote/GroupRemoteDataSource.kt [hunk 21]
```diff
@@ -151,28 +155,132 @@ class GroupRemoteDataSource : GroupDataSource {
...
+    }
+
+    override fun deleteCurrentRequestOfUserFromGroup(userId: String, request: Invite): Single<Boolean> {
+        TODO("Init later")
+    }
+
+    override fun createGroup(groupName: String, ownerId: String): Single<Boolean> {
+        // TODO: 06/12/2017
+        // I will optimize this function when I understand When-Then-And operator of Rx.
+        // at-cuongcao
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/way/data/source/remote/GroupRemoteDataSource.kt]: remove this line

**Comment by at-vinhhuynh** [comment]
@at-tienhoang 
I finish review this PR, please check it again.
thank you!

**Pull request merged**